source/common/http/http1/codec_impl.cc

```diff
@@ -119,6 +119,11 @@ void StreamEncoderImpl::encodeHeaders(const HeaderMap& headers, bool end_stream)
   if (end_stream) {
     endEncode();
   } else {
+    // Request is likely to be incomplete, in this case need to transfer the head request state into
```

Comment:
Sorry, I'm not clear on why this is necessary.  Our existing HEAD processing correctly determines when the headers are processed that the encode is complete and the new code path using sendLocalReply sets end stream explicitly.  Why do we need to do it manually in this new code path?

Reply:
If  this is an  incomplete request (end_stream=false), then `connection_.onEncodeComplete();` will not be called,  the head request state will not be passed to the corresponding response, causing http_parser to parse the response. "protocol error: HTTP / 1.1 protocol error: HPE_INVALID_EOF_STATE" appears, there is no way to make "expect true" for response->complete() in IdleTimeoutIntegrationTest

Reply:
Gotcha.   I think what this functionally does though is disallow HEAD requests with bodies.  I'd be very happy to disallow HEAD requests with bodies, but if we're going to do that I believe we should disallow GETs with bodies.  Either I think would be a higher risk change and I think should land as a separate PR :-)

If the problem is getting the information of "this is a response to the head request" to the response earlier, could we call
'pending_responses_.back().head_request_ = request_encoder_->headRequest();`
earlier in the chain, say when we've received the request headers?   @mattklein123 for attention on this one since it involves codec internals fairly deeply.

Reply:
> Gotcha. I think what this functionally does though is disallow HEAD requests with bodies. I'd be very happy to disallow HEAD requests with bodies, but if we're going to do that I believe we should disallow GETs with bodies. Either I think would be a higher risk change and I think should land as a separate PR :-)

Does it? I'm not sure it does?

> If the problem is getting the information of "this is a response to the head request" to the response earlier, could we call
'pending_responses_.back().head_request_ = request_encoder_->headRequest();`
earlier in the chain, say when we've received the request headers? @mattklein123 for attention on this one since it involves codec internals fairly deeply.

Yes, +1, we should do this outside of this function, and optimally only on the client side when headers are encoded. I think the optimal thing to do here would be to make a `ClientStreamEncoder` and override encodeHeaders() and then do this at the beginning of the call, before calling the base class encodeHeaders(). This is a larger change but I think would be more clear. Thoughts?

Reply:
Yeah, that'd work for me.   Looking at onEncodeComplete the only thing it does is push the state of head request, so rather than call onEncodeComplete when we're not done encodnig (end stream == false) I'm all for doing it when we get the headers.

Thanks for bearing with us, @zyfjeff while we sort this out :-)